frontend/src/pages/CompareV2.tsx

```diff
@@ -14,13 +14,166 @@
+import { ApiRunDetail } from 'src/apis/run';
+import Hr from 'src/atoms/Hr';
+import Separator from 'src/atoms/Separator';
+import CollapseButton from 'src/components/CollapseButton';
+import { QUERY_PARAMS, RoutePage } from 'src/components/Router';
+import { commonCss, padding } from 'src/Css';
+import { Apis } from 'src/lib/Apis';
+import Buttons from 'src/lib/Buttons';
+import { URLParser } from 'src/lib/URLParser';
+import { errorToMessage } from 'src/lib/Utils';
+import { classes, stylesheet } from 'typestyle';
+import { PageProps } from './Page';
+import RunList from './RunList';
+
+const css = stylesheet({
+  outputsRow: {
+    marginLeft: 15,
+    overflowX: 'auto',
+  },
+});
+
+export const overviewSectionName = 'Run overview';
+export const paramsSectionName = 'Parameters';
+export const metricsSectionName = 'Metrics';
+
+function CompareV2(props: PageProps) {
+  const { updateBanner, updateToolbar } = props;
+
+  const runlistRef = useRef<RunList>(null);
+  const [collapseSections, setCollapseSections] = useState<{ [key: string]: boolean }>({});
+  const [selectedIds, setSelectedIds] = useState<string[]>([]);
+
+  const queryParamRunIds = new URLParser(props).get(QUERY_PARAMS.runlist);
+  const runIds = (queryParamRunIds && queryParamRunIds.split(',')) || [];
+
+  // Retrieves run details.
+  const { isError, data, refetch } = useQuery<ApiRunDetail[], Error>(
+    ['run_details', { ids: runIds }],
+    () => Promise.all(runIds.map(async id => await Apis.runServiceApi.getRun(id))),
+    {
+      staleTime: Infinity,
+      onError: async error => {
+        const errorMessage = await errorToMessage(error);
+        updateBanner({
+          additionalInfo: errorMessage ? errorMessage : undefined,
+          message: `Error: failed loading ${runIds.length} runs. Click Details for more information.`,
+          mode: 'error',
+        });
+      },
+      onSuccess: () => props.updateBanner({}),
+    },
+  );
+
+  useEffect(() => {
+    const refresh = async () => {
+      if (runlistRef.current) {
+        await runlistRef.current.refresh();
+      }
+      await refetch();
+    };
+
+    const buttons = new Buttons(props, refresh);
+    updateToolbar({
+      actions: buttons
+        .expandSections(() => setCollapseSections({}))
+        .collapseSections(() =>
+          setCollapseSections({
+            [overviewSectionName]: true,
+            [paramsSectionName]: true,
+            [metricsSectionName]: true,
+          }),
+        )
+        .refresh(refresh)
+        .getToolbarActionMap(),
+      breadcrumbs: [{ displayName: 'Experiments', href: RoutePage.EXPERIMENTS }],
+      pageTitle: 'Compare runs',
+    });
+  }, [updateToolbar, refetch]);
+
+  useEffect(() => {
+    if (data) {
+      setSelectedIds(data.map(r => r.run!.id!));
+    }
+  }, [data]);
+
+  const collapseSectionsUpdate = (collapseSections: { [key: string]: boolean }): void => {
+    setCollapseSections({ ...collapseSections });
```

Comment:
I added the `{...}` destructuring function because React detects changes using [compare by reference](https://blog.bitsrc.io/understanding-referential-equality-in-react-a8fb3769be0), so in order for the page to detect the new `collapseSections` value and re-render, the variable must point to a new memory location.

Reply:
You are right: If we keep all collapseSection booleans in a single state variable, we will have to update the memory location for a re-render. However, since the number of sections is fixed (3 sections), how do you think about having three independent state variables? Hope it simplifies the logic a bit:

```
const [isOverviewCollapsed, setIsOverviewCollapsed] = useState(false);
const [isParamsCollapsed, setIsParamsCollapsed] = useState(false);
const [isMetricsCollapsed, setIsMetricsCollapsed] = useState(false);

...


return 
      ....
      <CollapseButton
        sectionName={OVERVIEW_SECTION_NAME}
        collapseSections={isOverviewCollapsed}
        collapseSectionsUpdate={setIsOverviewCollapsed}
      />
      ...
```

Note:
1. In this way, you don't need to define a `collapseSectionsUpdate` function for `<CollapseButton>`.
2. I see that `<CollapseButton>` doesn't have the same props as my example here. Feel free to create a new CollapseButton with the updated props definition. The main reason is that I don't think individual `CollapseButton` requires a whole `collapseSections` map, it only need to manipulate one boolean.
3. You might need to split one `setCollapseSections` into individual `setXXXX`'s during the collapse action in `updateToolbar()` call. But it should batch re-render for you as long as it is not async function.

Reply:
I can see how that is a bit simpler, though the downside is perhaps defining more variables and functions (variable for each section, new `CollapseButton` function. Right now, that button implementation is only used elsewhere in the `CompareV1` page, so changing the implementation is reasonable and we could phase out the old `CollapseButton` after V1.

However, I could address #1 above by changing the line in `CollapseButton` https://github.com/kubeflow/pipelines/blob/e3fc7cd365c23c810a9199f90715d2a140bd68a7/frontend/src/components/CollapseButton.tsx#L54
to the following:
```
collapseSectionsUpdate({...collapseSections});
```
Then I could directly pass in `setCollapseSections` to the `CollapseButton` in V2, and the V1 implementation would still work.

Do you think I should still create this new individual `CollapseButton` implementation? Happy to add that, just want to double-check the change doesn't unnecessarily add too much.

Reply:
Ah, actually looking further I just realized that the reason `CompareV1.tsx` designed the `collapseSections` this way is because new sections were dynamically added.

https://github.com/kubeflow/pipelines/blob/e3fc7cd365c23c810a9199f90715d2a140bd68a7/frontend/src/pages/CompareV1.tsx#L165-L205

I'll implement the changes as described, on further reflection I think that simplification makes sense.

Reply:
Thank you for the research! Sounds good on the new change.